chia/data_layer/data_store.py

```diff
@@ -843,6 +865,98 @@ async def delete(
 
         return
 
+    async def insert_batch(
+        self,
+        tree_id: bytes32,
+        changelist: List[Dict[str, Any]],
+        status: Status = Status.PENDING,
+        *,
+        lock: bool = True,
+    ) -> None:
+        async with self.db_wrapper.locked_transaction(lock=lock):
+            hint_keys_values = await self.get_keys_values_dict(tree_id, lock=False)
+            old_root = await self.get_tree_root(tree_id, lock=False)
+            for change in changelist:
+                if change["action"] == "insert":
+                    key = change["key"]
+                    value = change["value"]
+                    reference_node_hash = change.get("reference_node_hash")
+                    side = change.get("side")
+                    if reference_node_hash or side:
+                        await self.insert(key, value, tree_id, reference_node_hash, side, hint_keys_values, lock=False)
+                    else:
+                        await self.autoinsert(key, value, tree_id, hint_keys_values, lock=False)
+                else:
+                    assert change["action"] == "delete"
+                    key = change["key"]
+                    await self.delete(key, tree_id, hint_keys_values, lock=False)
+
+            root = await self.get_tree_root(tree_id, lock=False)
+            # We delete all "temporary" records stored in root and ancestor tables and store only the final result.
+            await self.rollback_to_generation(tree_id, old_root.generation, lock=False)
```

Comment:
yes - but only for the writer. the clients will download only the path from the root. the fix requires some extra work: traverse all intermediate roots, store all nodes, and then delete the nodes that are not on the path from the final root. I'm not sure this is worth it since all those tree crossings and deletions will contribute to the runtime of `batch_update`.